cdc/redo/writer/writer.go

```diff
@@ -0,0 +1,478 @@
+			}
+			logWriter.storage = s3storage
+		}
+
+		go logWriter.runGC(ctx)
+	})
+
+	return logWriter
+}
+
+func (l *LogWriter) initMeta(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return errors.Trace(ctx.Err())
+	default:
+	}
+
+	l.meta = &common.LogMeta{ResolvedTsList: map[int64]uint64{}}
+	files, err := ioutil.ReadDir(l.cfg.Dir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return cerror.WrapError(cerror.ErrRedoFileOp, errors.Annotate(err, "can't read log file directory"))
+	}
+
+	for _, file := range files {
+		if filepath.Ext(file.Name()) == common.MetaEXT {
+			path := filepath.Join(l.cfg.Dir, file.Name())
+			fileData, err := os.ReadFile(path)
+			if err != nil {
+				return cerror.WrapError(cerror.ErrRedoFileOp, err)
+			}
+
+			_, err = l.meta.UnmarshalMsg(fileData)
+			if err != nil {
+				l.meta = &common.LogMeta{ResolvedTsList: map[int64]uint64{}}
+				return cerror.WrapError(cerror.ErrRedoFileOp, err)
+			}
+			break
+		}
+	}
+	return nil
+}
+
+func (l *LogWriter) runGC(ctx context.Context) {
+	ticker := time.NewTicker(time.Duration(defaultGCIntervalInMs) * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		if l.isStopped() {
+			return
+		}
+
+		select {
+		case <-ctx.Done():
+			log.Info("runGC got canceled", zap.Error(ctx.Err()))
+			return
+		case <-ticker.C:
+			err := l.gc()
+			if err != nil {
+				log.Error("redo log GC error", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (l *LogWriter) gc() error {
+	l.metaLock.RLock()
+	defer l.metaLock.RUnlock()
+
+	var err error
+	ts := l.meta.CheckPointTs
+	err = multierr.Append(err, l.rowWriter.GC(ts))
+	err = multierr.Append(err, l.ddlWriter.GC(ts))
+	return err
+}
+
+// WriteLog implement WriteLog api
+func (l *LogWriter) WriteLog(ctx context.Context, tableID int64, rows []*model.RedoRowChangedEvent) (uint64, error) {
+	select {
+	case <-ctx.Done():
+		return 0, errors.Trace(ctx.Err())
+	default:
+	}
+
+	if l.isStopped() {
+		return 0, cerror.ErrRedoWriterStopped
+	}
+	if len(rows) == 0 {
+		return 0, nil
+	}
+
+	maxCommitTs := l.setMaxCommitTs(tableID, 0)
+	for _, r := range rows {
+		if r == nil || r.Row == nil {
+			continue
+		}
+
+		rl := redoLogPool.Get().(*model.RedoLog)
+		rl.Row = r
+		rl.DDL = nil
+		rl.Type = model.RedoLogTypeRow
+		// TODO: crc check
+		data, err := rl.MarshalMsg(nil)
+		if err != nil {
+			// TODO: just return 0 if err ?
+			return maxCommitTs, cerror.WrapError(cerror.ErrMarshalFailed, err)
+		}
+
+		l.rowWriter.AdvanceTs(r.Row.CommitTs)
+		_, err = l.rowWriter.Write(data)
+		if err != nil {
+			return maxCommitTs, err
+		}
+		maxCommitTs = l.setMaxCommitTs(tableID, r.Row.CommitTs)
+		redoLogPool.Put(rl)
+	}
+	return maxCommitTs, nil
+}
+
+// SendDDL implement SendDDL api
+func (l *LogWriter) SendDDL(ctx context.Context, ddl *model.RedoDDLEvent) error {
+	select {
+	case <-ctx.Done():
+		return errors.Trace(ctx.Err())
+	default:
+	}
+
+	if l.isStopped() {
+		return cerror.ErrRedoWriterStopped
```

Comment:
```suggestion
		return cerror.ErrRedoWriterStopped.GenWithStackByArgs()
```
ditto for other rfc error usage.